composer/core/evaluator.py

```diff
@@ -39,6 +41,23 @@ def should_eval(state: State, event: Event):
         if int(eval_interval) <= 0:
             return False
 
+        # if requested, evaluate at the end of training, as long as the length of training is specified.
+        if eval_at_fit_end and state.max_duration is not None:
+            # check if we are at the end of training
+            # handle different eval_intervals in order to not duplicate validation steps
+            if eval_interval.unit == TimeUnit.EPOCH and event == Event.EPOCH_END:
```

Comment:
Could this logic be refactored to work if `max_duration` is specified in tokens or samples? (In addition, I'd prefer to avoid relying on `dataloader_len`, as it can be an approximation and is not always defined.)

I think when using `Event.FIT_END` the logic could be something like:
```python

last_eval_batch = None

def should_eval(state, event):
  nonlocal last_eval_batch  # Not sure if this line is needed? It might be
  
  ... # existing eval logic. Make sure to update last_eval_batch appropriately.
  if event == Event.FIT_END and state.timestamp.batch != last_eval_batch and eval_at_fit_end:
    last_eval_batch = state.timestamp.batch
    return True
```

Reply:
(Implemented)